core/env.go

```diff
@@ -153,12 +154,84 @@ func (env *Env) Ingest(obj dagql.Object) string {
 	if !ok {
 		env.typeCount[typeName]++
 		llmID = fmt.Sprintf("%s#%d", typeName, env.typeCount[typeName])
+		if desc == "" {
+			desc = env.describe(obj.ID())
+		}
 		env.idByHash[hash] = llmID
-		env.objsByID[llmID] = &Binding{Key: llmID, Value: obj, env: env}
+		env.objsByID[llmID] = &Binding{
+			Key:         llmID,
+			Value:       obj,
+			Description: desc,
+			env:         env,
+		}
 	}
 	return llmID
 }
 
+func (env *Env) describe(id *call.ID) string {
+	str := new(strings.Builder)
+	if recv := id.Receiver(); recv != nil {
+		if llmID, ok := env.idByHash[recv.Digest()]; ok {
+			str.WriteString(llmID)
+		} else {
+			str.WriteString(recv.Digest().String())
+		}
+		str.WriteString(".")
+	}
+	str.WriteString(id.Field())
+
+	// Include arguments in the description
+	if args := id.Args(); len(args) > 0 {
+		str.WriteString("(")
+		for i, arg := range args {
+			if i > 0 {
+				str.WriteString(", ")
+			}
+			str.WriteString(arg.Name())
+			str.WriteString(": ")
+			str.WriteString(env.displayLit(arg.Value()))
+		}
+		str.WriteString(")")
+	}
+	return str.String()
+}
+
+func (env *Env) displayLit(lit call.Literal) string {
+	switch x := lit.(type) {
+	case *call.LiteralID:
+		// For ID arguments, try to use LLM IDs
+		if llmID, ok := env.idByHash[x.Value().Digest()]; ok {
+			return llmID
+		} else {
+			return x.Value().Type().NamedType()
+		}
+	case *call.LiteralList:
+		list := "["
+		_ = x.Range(func(i int, value call.Literal) error {
+			if i > 0 {
+				list += ","
+			}
+			list += env.displayLit(value)
+			return nil
+		})
+		list += "]"
+		return list
+	case *call.LiteralObject:
+		obj := "{"
+		_ = x.Range(func(i int, name string, value call.Literal) error {
+			if i > 0 {
+				obj += ","
+			}
+			obj += name + ": " + env.displayLit(value)
+			return nil
+		})
+		obj += "}"
+		return obj
+	default:
+		return lit.Display()
+	}
+}
+
 func (env *Env) Types() []string {
 	types := make([]string, 0, len(env.typeCount))
 	for typ := range env.typeCount {
```

Comment:
nit: i initially thought this was a `range int`, like for `typ := range 10` and typ was incrementing integers lmao
```suggestion
	for typ := range env.typeCounts {
```